File: ts/src/scopes.test.ts
import { describe, expect, it } from "vitest";

import { Limiter } from "./limiter";
import { LimitRule, Scope } from "./rules";

// Hierarchical scopes: global fleet, per-tenant, per-user - all enforced together.

describe("scopes", () => {
  it("isolates tenants under a tenant-scoped rule", () => {
    const l = new Limiter([new LimitRule("gpt-4o", 60, { maxRequests: 2, scope: Scope.Tenant })]);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", now: 0 }).allowed).toBe(true);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", now: 0 }).allowed).toBe(true);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", now: 0 }).allowed).toBe(false);
    // A different tenant has its own budget.
    expect(l.tryAcquire("gpt-4o", { tenant: "globex", now: 0 }).allowed).toBe(true);
  });

  it("isolates users within a tenant under a user-scoped rule", () => {
    const l = new Limiter([new LimitRule("gpt-4o", 60, { maxRequests: 1, scope: Scope.User })]);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", user: "ann", now: 0 }).allowed).toBe(true);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", user: "ann", now: 0 }).allowed).toBe(false);
    // Same tenant, different user.
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", user: "bob", now: 0 }).allowed).toBe(true);
  });

  it("shares a global-scoped rule across everyone", () => {
    const l = new Limiter([new LimitRule("gpt-4o", 60, { maxRequests: 2, scope: Scope.Global })]);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", user: "ann", now: 0 }).allowed).toBe(true);
    expect(l.tryAcquire("gpt-4o", { tenant: "globex", user: "bob", now: 0 }).allowed).toBe(true);
    // The fleet cap is hit no matter who is calling.
    expect(l.tryAcquire("gpt-4o", { tenant: "new", user: "cat", now: 0 }).allowed).toBe(false);
  });

  it("lets the toughest applicable rule win", () => {
    // A generous per-tenant cap but a tiny global fleet cap.
    const l = new Limiter([
      new LimitRule("gpt-4o", 60, { maxRequests: 1000, scope: Scope.Tenant, name: "tenant" }),
      new LimitRule("gpt-4o", 60, { maxRequests: 1, scope: Scope.Global, name: "fleet" }),
    ]);
    expect(l.tryAcquire("gpt-4o", { tenant: "acme", now: 0 }).allowed).toBe(true);
    const d = l.tryAcquire("gpt-4o", { tenant: "acme", now: 0 });
    expect(d.allowed).toBe(false);
    expect(d.trippedRule?.name).toBe("fleet");
    expect(d.scope).toBe(Scope.Global);
  });
});

File: go/scopes_test.go
package quotagate

import "testing"

// Hierarchical scopes: global fleet, per-tenant, per-user - all enforced together.

func TestTenantScopeIsolatesTenants(t *testing.T) {
	l := NewLimiter([]LimitRule{NewLimitRule("gpt-4o", 60, WithMaxRequests(2), WithScope(Tenant))})
	if !l.TryAcquire("gpt-4o", WithTenant("acme"), WithNow(0)).Allowed {
		t.Fatal("acme call 1 should be allowed")
	}
	if !l.TryAcquire("gpt-4o", WithTenant("acme"), WithNow(0)).Allowed {
		t.Fatal("acme call 2 should be allowed")
	}
	if l.TryAcquire("gpt-4o", WithTenant("acme"), WithNow(0)).Allowed {
		t.Fatal("acme is now full")
	}
	// A different tenant has its own budget.
	if !l.TryAcquire("gpt-4o", WithTenant("globex"), WithNow(0)).Allowed {
		t.Fatal("globex should have its own budget")
	}
}

func TestUserScopeIsolatesUsersWithinATenant(t *testing.T) {
	l := NewLimiter([]LimitRule{NewLimitRule("gpt-4o", 60, WithMaxRequests(1), WithScope(User))})
	if !l.TryAcquire("gpt-4o", WithTenant("acme"), WithUser("ann"), WithNow(0)).Allowed {
		t.Fatal("ann call 1 should be allowed")
	}
	if l.TryAcquire("gpt-4o", WithTenant("acme"), WithUser("ann"), WithNow(0)).Allowed {
		t.Fatal("ann is now full")
	}
	// Same tenant, different user.
	if !l.TryAcquire("gpt-4o", WithTenant("acme"), WithUser("bob"), WithNow(0)).Allowed {
		t.Fatal("bob should have his own budget")
	}
}

func TestGlobalScopeIsSharedAcrossEveryone(t *testing.T) {
	l := NewLimiter([]LimitRule{NewLimitRule("gpt-4o", 60, WithMaxRequests(2), WithScope(Global))})
	if !l.TryAcquire("gpt-4o", WithTenant("acme"), WithUser("ann"), WithNow(0)).Allowed {
		t.Fatal("first global call should be allowed")
	}
	if !l.TryAcquire("gpt-4o", WithTenant("globex"), WithUser("bob"), WithNow(0)).Allowed {
		t.Fatal("second global call should be allowed")
	}
	// The fleet cap is hit no matter who is calling.
	if l.TryAcquire("gpt-4o", WithTenant("new"), WithUser("cat"), WithNow(0)).Allowed {
		t.Fatal("the fleet cap should be hit regardless of caller")
	}
}

func TestToughestApplicableRuleWins(t *testing.T) {
	// A generous per-tenant cap but a tiny global fleet cap.
	l := NewLimiter([]LimitRule{
		NewLimitRule("gpt-4o", 60, WithMaxRequests(1000), WithScope(Tenant), WithName("tenant")),
		NewLimitRule("gpt-4o", 60, WithMaxRequests(1), WithScope(Global), WithName("fleet")),
	})
	if !l.TryAcquire("gpt-4o", WithTenant("acme"), WithNow(0)).Allowed {
		t.Fatal("first call should be allowed")
	}
	d := l.TryAcquire("gpt-4o", WithTenant("acme"), WithNow(0))
	if d.Allowed {
		t.Fatal("the tiny fleet cap should trip")
	}
	if d.TrippedRule == nil || d.TrippedRule.Name != "fleet" {
		t.Fatalf("expected fleet tripped, got %+v", d.TrippedRule)
	}
	if d.Scope == nil || *d.Scope != Global {
		t.Fatalf("expected global scope, got %v", d.Scope)
	}
}
